### PR TITLE
Blueprint `PlanningInput` API rework

### PR DESCRIPTION
**Pull request by jgallagher**
The impetus for this change was [@davepacheco's comment on #5344](https://github.com/oxidecomputer/omicron/pull/5344#discussion_r1544638094). Summarizing the changes, several of which are open to bikeshedding:

* The real meat of the PR: None of `PlanningInput`'s fields are public anymore; all access have to go through methods, and none of the methods allow mutation. The methods that iterate over sleds take a required `SledFilter`; making use of this revealed at least one legit bug (prior to this PR, the planner would attempt to put an NTP zone on a sled if it was in the `(SledPolicy::Expunged, SledState::Active)` state and had no NTP zone, because it was checking `SledState` but not `SledPolicy`).
* There is a separate `PlanningInputBuilder` that allows mutation; all the unit tests go through this. For tests that want to take a `PlanningInput` and modify it, they use the pattern `PlanningInput::into_builder()` -> mutate via builder as desired -> `PlanningInputBuilder::build()`.
* I used a `TypedUuid<SledKind>` in `PlanningInput{,Builder}`. I did not change all the other sled IDs, so this diff is littered with `TODO-cleanup`s where I'm having to convert between typed and untyped. This is already big enough to review, so my preference is to do this in a followup PR.
* `Policy` no longer includes the `sleds`; it's conceptually more like "system-level policy". Currently it only includes the service IP pool ranges and the target Nexus count.
* I broke `SledResources` up a bit, removing the sled state and policy. The resources+policy+state are combined in `SledDetails`. (This was just for clarity, because seeing "policy" under "resources" seems confusing, but if I'm in the minority it'd be easy to change this back.)
* Internal and external DNS versions are now embedded in `PlanningInput` instead of being separate arguments to `Planner::new_based_on()`. They _are_ still separate arguments to `BlueprintBuilder::build_initial_from_collection()`, because that doesn't take a `PlanningInput` (but it still wants the DNS versions). I think this is fine, especially since we plan to drop `build_initial_from_collection()` in the next few weeks anyway.

I think this is worth doing for the `SledFilter`-makes-it-harder-to-write-bugs alone, but hopefully it also hits some of Dave's goals with reducing churn as we modify `PlanningInput` in the future.

### CODE REVIEW

File: nexus/types/src/deployment/planning_input.rs [hunk 1]
```diff
@@ -0,0 +1,397 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+//! Types describing inputs the Reconfigurator needs to plan and produce new
```
Comment (andrewjstone) [inline nexus/types/src/deployment/planning_input.rs:5]: It's very nice to have this in a separate, nicely labeled file!

**Review by andrewjstone — APPROVED** [review]
Looks great. Excellent cleanup @jgallagher.

**Review by sunshowers — APPROVED** [review]
Thanks!

File: dev-tools/reconfigurator-cli/src/main.rs [hunk 2]
```diff
@@ -381,7 +411,7 @@ struct SledAddArgs {
 #[derive(Debug, Args)]
 struct SledArgs {
     /// id of the sled
-    sled_id: Uuid,
+    sled_id: TypedUuid<SledKind>,
```
[the remark this answers is not in the document]
Comment (sunshowers) [reply dev-tools/reconfigurator-cli/src/main.rs:414]: (also we can slowly migrate over this code as well. this is so exciting)

File: dev-tools/reconfigurator-cli/src/main.rs [hunk 3]
```diff
@@ -111,15 +113,45 @@ impl ReconfiguratorSim {
...
+        //   executing and newer if other blueprints have changed DNS in the
+        //   meantime).
+        //
+        // But in this CLI, there's no execution at all.  As a result, there's
+        // no way to really choose between these -- and it doesn't really
+        // matter, either.  We'll just pick the parent blueprint's.
+        *builder.internal_dns_version_mut() =
+            parent_blueprint.internal_dns_version;
+        *builder.external_dns_version_mut() =
+            parent_blueprint.external_dns_version;
```
[the remark this answers is not in the document]
Comment (jgallagher) [reply dev-tools/reconfigurator-cli/src/main.rs:149]: Yeah, that would be better. Will fix.

**Pull request merged**